### PR TITLE
sync: fix missing notification during mpsc close

### PR DESCRIPTION
**Pull request by carllerche**
When the mpsc channel receiver closes the channel, receiving should
return `None` once all in-progress sends have completed. When a sender
reserves capacity, this prevents the receiver from fully shutting down.
Previously, when the sender, after reserving capacity, dropped without
sending a message, the receiver was not notified. This results in
blocking the shutdown process until all sender handles drop.

This patch adds a receiver notification when the channel is both closed
and all outstanding sends have completed.

### CODE REVIEW

File: tokio/src/sync/mpsc/chan.rs [hunk 1]
```diff
@@ -75,10 +75,15 @@ pub(crate) trait Semaphore {
...
+    ///
+    /// # Return
+    ///
+    /// Returns true if the permit was acquired.
+    fn drop_permit(&self, permit: &mut Self::Permit) -> bool;
 
     fn is_idle(&self) -> bool;
 
+    // Returns `true` if the receiver should be notified for shutdown.
     fn add_permit(&self);
```
Comment (Darksonn) [inline tokio/src/sync/mpsc/chan.rs:87]: This comment appears to be wrong.

File: tokio/src/sync/mpsc/chan.rs [hunk 2]
```diff
@@ -234,7 +239,11 @@ where
     S: Semaphore,
 {
     fn drop(&mut self) {
-        self.inner.semaphore.drop_permit(&mut self.permit);
+        let notify = self.inner.semaphore.drop_permit(&mut self.permit);
+
+        if notify && self.inner.semaphore.is_idle() {
```
Comment (Darksonn) [inline tokio/src/sync/mpsc/chan.rs:244]: Does `is_idle` mean that no permits are currently taken?

Comment (carllerche) [reply tokio/src/sync/mpsc/chan.rs:244]: Correct

**Comment by carllerche** [comment]
As an additional note, I am in the process of changing *most* of this code for 0.3.

**Review by Darksonn — APPROVED** [review]
It looks correct to me.

**Review by hawkw — APPROVED** [review]
modulo alice's suggestions, this looks good to me.

**Pull request merged**